<!-- DO NOT MODIFY README.md!!! It was automatically generated by GitHub Actions and will be overwritten on PR merge! -->

John Gaspar
- 19 Jul, 2019

Scott Norton
- 12 Jul, 2022
- 01 Jul, 2024
- 13 Sep, 2024
- 22 Jan, 2025

<a name="intro"></a>

## Introduction

DRUG-seq (**D**igital **R**NA with pert**U**rbation of **G**enes) [[1](#references)] is a new technique that provides a cost-effective way for high-throughput, whole transcriptome RNA-seq profiling.  Cells grown in multiwell plates are lysed, and reverse transcription is performed using oligos that contain barcodes that are unique to each well ([Fig. 1a](https://github.com/MSDLLCPapers/DRAGoN/blob/master/figures/figure1.jpg), from [Ye *et al.*, 2018](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6192987/)).  After paired-end sequencing, the R1 reads contain the barcodes, as well as unique molecular identifiers (UMIs) that allow for the removal of PCR duplicates ([Fig. 1b](https://github.com/MSDLLCPapers/DRAGoN/blob/master/figures/figure1.jpg)).  The R2 reads match the original mRNA sequence.

The computational pipeline consists of several steps: aligning the R2 reads to a genome (or transcriptome), counting gene overlaps, and then demultiplexing and removing PCR duplicates using the R1 reads.  The analyses in the DRUG-seq paper were performed with custom code, to which we do not have access.  Instead, we modified the SpatialTranscriptomics Pipeline [[2](#references)] to produce output matrices for DRUG-seq data ([Fig. 2](https://github.com/MSDLLCPapers/DRAGoN/blob/master/figures/figure2.png), from [Fern&aacute;ndez Navarro *et al.*, 2017](https://academic.oup.com/bioinformatics/article/33/16/2591/3111847)). This pipeline, based on [ST Pipeline v1.7.6](https://github.com/SpatialTranscriptomicsResearch/st_pipeline/releases/tag/1.7.6), would time out on production-scale datasets, necessitating a new approach. Attempts to co-opt zUMIs [[4](#references)] allowed a modest improvement in the maximum dataset size we could handle, however this would fail on memory constraints.

This repository presents two alternative pipelines implemented in nextflow. The first, `STARsolo.nf`, wraps STAR running in Solo mode [[5](#references)] to provide a rough estimate of UMI counts. The second is `DRAGoN.nf`, a novel proprietary pipeline called **D**RUG-seq/**R**NA-seq **A**nalysis of **G**enetically-perturbed cells **o**n **N**extflow (**DRAGoN**), employs a more robust duplicate removal algorithm. Both of these pipelines leverage Nextflow's efficient resource distribution to accelerate processing while keeping memory usage in check.
